### PR TITLE
chore: refactor plan to be more maintainable

### PR DESCRIPTION
**Pull request by tobymao**

### CODE REVIEW

File: tests/core/test_plan.py [hunk 1]
```diff
@@ -1548,7 +1548,7 @@ def test_indirectly_modified_forward_only_model(make_snapshot, mocker: MockerFix
 
     assert updated_snapshot_a.change_category == SnapshotChangeCategory.BREAKING
     assert updated_snapshot_b.change_category == SnapshotChangeCategory.FORWARD_ONLY
-    assert updated_snapshot_c.change_category == SnapshotChangeCategory.INDIRECT_BREAKING
+    assert updated_snapshot_c.change_category == SnapshotChangeCategory.FORWARD_ONLY
```
Comment (izeigerman) [inline tests/core/test_plan.py:1551]: Let's also make test in which C points at both A and B, in which case, the C's category should indeed be `INDIRECT_BREAKING`.

File: sqlmesh/core/plan/builder.py [hunk 2]
```diff
@@ -180,9 +181,15 @@ def set_choice(self, snapshot: Snapshot, choice: SnapshotChangeCategory) -> Plan
             snapshot: The target snapshot.
             choice: The user decision on how to version the target snapshot and its children.
         """
-        plan = self.build()
-        self._set_choice(snapshot, choice, plan.directly_modified, plan.indirectly_modified)
-        self._adjust_new_snapshot_intervals()
+        if self._forward_only:
```
Comment (izeigerman) [inline sqlmesh/core/plan/builder.py:184]: Do we also need to check that the category is not assigned to an indirectly modified snapshot?

Comment (tobymao) [reply sqlmesh/core/plan/builder.py:184]: i don't understand this comment, can you explain it to me?

File: tests/core/test_plan.py (hunk 1 above)
Comment (tobymao) [reply tests/core/test_plan.py]: done

File: sqlmesh/utils/dag.py [hunk 3]
```diff
@@ -30,7 +31,8 @@ def add(self, node: T, dependencies: t.Optional[t.Iterable[T]] = None) -> None:
             node: The node to add.
             dependencies: Optional dependencies to add to the node.
         """
-        self._sorted = None
+        self._sorted = []
```
Comment (izeigerman) [inline sqlmesh/utils/dag.py:34]: IMHO I preferred the explicit `None` way. This way we have a clear delineation between an empty DAG and missing cache.

File: sqlmesh/utils/dag.py [hunk 4]
```diff
@@ -30,7 +31,8 @@ def add(self, node: T, dependencies: t.Optional[t.Iterable[T]] = None) -> None:
...
+        self._upstream = {}
```
Comment (izeigerman) [inline sqlmesh/utils/dag.py:35]: Ditto

**Pull request merged**